File: solutions/letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.go
// @Title: Letter Combinations of a Phone Number
// @Author: colinjxc
// @Date: 2021-05-23T22:38:32+08:00
// @URL: https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number


var (
	letterMap = []string{
		" ",    //0
		"",     //1
		"abc",  //2
		"def",  //3
		"ghi",  //4
		"jkl",  //5
		"mno",  //6
		"pqrs", //7
		"tuv",  //8
		"wxyz", //9
	}
)

func findCombinations(digits *string, index int, combine string, res *[]string) {
	if len(*digits) == index {
		*res = append(*res, combine)
		return
	}
	digit := (*digits)[index] - '0'
	digitLetters := letterMap[digit]

	for i := 0; i < len(digitLetters); i++ {
		findCombinations(digits, index+1, combine+string(digitLetters[i]), res)
	}
}

func letterCombinations(digits string) []string {
	var res []string
	if len(digits) == 0 {
		return res
	}
	findCombinations(&digits, 0, "", &res)
	return res
}

File: solutions/letter-combinations-of-a-phone-number/letter-combinations-of-a-phone-number.cpp
// @Title: Letter Combinations of a Phone Number
// @Author: colinjxc
// @Date: 2022-02-08T20:04:11+08:00
// @URL: https://leetcode-cn.com/problems/letter-combinations-of-a-phone-number


class Solution {
private:
    string digitStr[10] = {"", "", "abc", "def", "ghi", 
                        "jkl", "mno", "pqrs", "tuv", "wxyz"};
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        dfs(0, digits, "", res);
        return res;
    }

    void dfs(int index, string& digits, const string& tmp, vector<string>& res) {
        if (digits.size() == 0) return;
        if (index == digits.size()) {
            res.emplace_back(tmp);
            return;
        }
        int num = digits[index] - '0';
        for (auto& ch : digitStr[num]) {
            dfs(index + 1, digits, tmp + ch, res);
        }
    }
};
